scripts/megastock/write_databricks_to_bigquery_03.py

```diff
@@ -18,52 +25,116 @@
+
+# set up paths to write to 
+bq_project = "cube-machine-learning"
+bq_dataset = "ds_api_datasets"
+bq_megastock_table = 'megastock_combined_baseline'
+bq_write_path = f"{bq_project}.{bq_dataset}.{bq_megastock_table}"
+
+# COMMAND ----------
+
+# read in data
 building_metadata = spark.table(f'ml.megastock.building_metadata_{N_SAMPLE_TAG}')
+building_features = spark.table(f'ml.megastock.building_features_{N_SAMPLE_TAG}')
+
+# COMMAND ----------
+
+# get distinct climate zones
+climate_zone_list = list(map(lambda row: row['ashrae_iecc_climate_zone_2004'], building_metadata.select('ashrae_iecc_climate_zone_2004').distinct().collect()))
 
+# Mapping dictionary
+climate_zone_mapping = {cz: i for i, cz in (enumerate(sorted(climate_zone_list), 1))}
+
+# Define UDF
+cz_mapping_expr = F.create_map([F.lit(x) for x in chain(*climate_zone_mapping.items())])
+
+# Apply UDF to create new column
+building_metadata_with_int_cz = building_metadata.withColumn("climate_zone_int", cz_mapping_expr[F.col('ashrae_iecc_climate_zone_2004')])
 
 # COMMAND ----------
 
-(building_metadata
-    .write
-    .format("bigquery")
-    .mode("overwrite")
-    .option("table", bq_metadata_write_path)
-    .option("temporaryGcsBucket", "the-cube")
-    .save()
+# Add "__m" suffix to all columns in metadata table except "building_id"
+building_metadata_renamed = building_metadata_with_int_cz.select(
+    [F.col(c).alias(f"{c}__m") if c != "building_id" else F.col(c) for c in building_metadata_with_int_cz.columns]
 )
 
 # COMMAND ----------
 
-bq_features_write_path = 'cube-machine-learning.ds_api_datasets.megastock_features'
-building_features = spark.table(f'ml.megastock.building_features_{N_SAMPLE_TAG}')
+# Subset features to baseline only and join the tables on building_id
+combined_df_baseline = building_metadata_renamed.join(
+    building_features.where(F.col('upgrade_id') == 0).drop('upgrade_id'),
+    on="building_id", how="inner")
+
+# COMMAND ----------
 
-(building_features
+# write out table to big query
+(combined_df_baseline
     .write
     .format("bigquery")
     .mode("overwrite")
-    .option("table", bq_features_write_path)
+    .option("table",bq_write_path)
     .option("temporaryGcsBucket", "the-cube")
     .save()
 )
 
-
 # COMMAND ----------
 
-# check that tables are there
-client = bigquery.Client()
+# Drop table
+query = f"""
+DROP TABLE `{bq_write_path}`"""
+query_job = client.query(query)
+query_job.result()
+
+# optimize the table by partitioning and clustering
+query = f"""
+CREATE TABLE `{bq_write_path}_optimized`
+PARTITION BY RANGE_BUCKET(climate_zone_int, GENERATE_ARRAY(1, {len(climate_zone_mapping)+1}, 1))
+CLUSTER BY  heating_fuel__m, geometry_building_type_acs__m, geometry_floor_area__m, vintage__m AS
+SELECT *,
+FROM `{bq_write_path}`
+"""
+
+# Run the query
+query_job = client.query(query)
+# Wait for the query to complete
+query_job.result()
+
+# Drop non-optimized table
+query = f"""
+DROP TABLE `{bq_write_path}`"""
```

Comment:
Tiny formatting nit:
```suggestion
query = f"""DROP TABLE `{bq_write_path}`"""
```